pyspark_datasources/restapi.py

```diff
@@ -0,0 +1,402 @@
+    def __init__(self, options):
+        super().__init__(options)
+        if "path" not in options or not options["path"]:
+            raise Exception("You must specify the URL to read from in `.load()`.")
+
+    @classmethod
+    def name(cls):
+        return "restapi"
+
+    def schema(self):
+        return "struct<url: string, headers: map<string, string>, auth: map<string, string>, params: map<string, string>, data: string, json_data: map<string, string>, method: string, options: map<string, string>, result: string>"
+
+    def reader(self, schema: T.StructType) -> DataSourceReader:
+        return RestapiRequestReader(self.options)
+
+class RestapiRequestReader(DataSourceReader):
+    def __init__(self, options: dict):
+        self.protocol = str(options.get("protocol", "https")).strip()
+        _ = self.valid_method(self.protocol)
+
+        _ = validators.url(options.get("path"))
+        self.url = f"{self.protocol}://{options.get('path')}"
+
+        self.headers = self._get_dict_option("headers", options) or {}
+
+        self.auth = self._get_dict_option("auth", options) or { "type": "NO_AUTH" }
+        self.auth["type"] = str(self.auth.get("type")).strip().upper()
+        _ = self.valid_auth(self.auth)
+
+        self.params = self._get_dict_option("params", options) or {}
+
+        self.data = str(options.get("data", "")).strip()
+        _ = self.valid_data(self.data)
+
+        self.json_data = self._get_dict_option("json_data", options) or {}
+        _ = self.valid_json_data(self.json_data)
+
+        self.method = str(options.get("method", "GET")).strip().upper()
+        _ = self.valid_method(self.method)
+
+        # For options, see: https://www.python-httpx.org/api
+        # Advanced use, they have the potential to overwrite the other exposed options
+        self.options = self._get_dict_option("options", options) or {}
+
+    def read(self, partition):
+        for row in self._batch_request():
+            yield Row(
+                url=self.url,
+                headers=self.headers,
+                auth=self.auth,
+                params=self.params,
+                data=self.data,
+                json_data=self.json_data,
+                method=self.method,
+                options=self.options,
+                result=row
+            )
+
+    def _get_dict_option(self, key: str, options: dict) -> dict:
+        """Get a dictionary option from the options. Spark automatically transforms them to string.
+
+        Parameters
+        ----------
+        key : str
+            Option key, as passed to `spark.read.option(key, value)`
+        options : dict
+            Options dictionary
+
+        Raises
+        ------
+            TypeError
+                Option value is expected to be of either type str or dict
+
+        Returns
+        -------
+        dict
+            Dictionary containing requested option
+        """
+        option_value = options.get(key)
```

Comment:
I was trying to use the data source (with specifying headers) but got this error: `TypeError: Expected headers to be either str or dict. Got <class 'NoneType'>`. I think we can return None if the option value is None.

Reply:
That was an oversight on my part, but should be fixed now